### PR TITLE
Add useful links when build fails

### PR DESCRIPTION
**Pull request by melix**
### Context

Whenever a build fails, it might not be super obvious for a user what to do. Either because they don't know Gradle, or because the error message is too obscure. For this reason, this pull request makes sure that whenever a build fails:

- a link to the free training registration page is shown
- a link to our forums is added
- a link to build scans is added if the build scan plugin wasn't applied

This PR adds those unconditionally, but we might want in the future to try to display a selection of links/hints depending on the context (is it a test failure? is it a build script compile error?)

### CODE REVIEW

**Comment by eriwen** [comment]
@melix is it your plan to address the "todo" to hide the build scan message if `--scan` is applied before merging this PR?

**Comment by melix** [comment]
No, as I think it requires more work than we can afford for this.

File: subprojects/core/src/main/java/org/gradle/internal/buildevents/BuildExceptionReporter.java [hunk 1]
```diff
@@ -176,28 +204,45 @@ private LinePrefixingStyledTextOutput getLinePrefixingStyledTextOutput() {
...
-            details.resolution.text(" option to get the stack trace. ");
+            resolution.text("Run with ");
+            maybeAddBuildScan(resolution);
+            resolution.withStyle(UserInput).format("--%s", LoggingCommandLineConverter.STACKTRACE_LONG);
+            resolution.text(" option to get the stack trace. ");
         }
         if (loggingConfiguration.getLogLevel() != LogLevel.DEBUG) {
-            details.resolution.text("Run with ");
+            resolution.text("Run with ");
+            maybeAddBuildScan(resolution);
```
Comment (eriwen) [inline subprojects/core/src/main/java/org/gradle/internal/buildevents/BuildExceptionReporter.java]: :x: Along with line 216, this prints the "Run with --scan" twice. 

screen shot 2017-07-10 at 9 33 11 am

**Comment by melix** [comment]
Fixed, and rebased onto `release`.

**Comment by ldaley** [comment]
Are we sure we want to release this with the awkward experience of using `--scan` when the plugin isn't applied, which is the case for most of our users?

**Comment by melix** [comment]
@ldaley that's a fair question, and my reasoning was that lots of actual users don't write the build files themselves, and don't follow our newsletters, releases, ... They are just consumers. As such, they ignore there is such a thing as build scans. So I agree the experience would be awkward, because you would do:

`./gradle doSomething`

fails, and tells you "run with `--scan`". Then you do `./gradlew --scan doSomething` and it would fail again, this time telling you "please apply the scan plugin".

One option is just to remove the mention to `--scan` until we have a release which automatically applies the plugin if `--scan` is used. The drawback is that we have no ETA for this, so we're potentially missing an opportunity to promote the scan plugin.

**Comment by pioterj** [comment]
Apart from missing help page, is this good to go?

**Comment by ldaley** [comment]
@melix yeah, it's a tough call. Movement on built in build scans will happen soon. For the time being, we should avoid the shitty user experience.

I vote for removing the --scan wording until we have a better answer. We can always add it in then.

**Comment by pioterj** [comment]
I agree with @ldaley. I'd postpone adding mention of `--scan` until they are built-in. 

@melix How does the output look like after all the changes? Could you add a screenshot?

**Comment by melix** [comment]
I removed the `--scan` hint. Here's what it looks like now: https://scans.gradle.com/s/utobgup25pucm/log

File: subprojects/core/src/main/java/org/gradle/internal/buildevents/BuildExceptionReporter.java [hunk 2]
```diff
@@ -49,13 +51,36 @@
...
 
+    private Gradle gradle;
+
     public BuildExceptionReporter(StyledTextOutputFactory textOutputFactory, LoggingConfiguration loggingConfiguration, BuildClientMetaData clientMetaData) {
         this.textOutputFactory = textOutputFactory;
         this.loggingConfiguration = loggingConfiguration;
         this.clientMetaData = clientMetaData;
     }
 
+    @Override
```
Comment (eriwen) [inline subprojects/core/src/main/java/org/gradle/internal/buildevents/BuildExceptionReporter.java]: Are these changes also unnecessary now?

Comment (melix) [inline subprojects/core/src/main/java/org/gradle/internal/buildevents/BuildExceptionReporter.java]: removed

**Comment by eriwen** [comment]
Thanks @melix!

**Pull request merged**